### ISSUE TITLE
Commands to vehicle sometimes appear to be lost

### ISSUE THREAD

**Issue by tfoote** [mavlink/qgroundcontrol#4232]
I'm seeing behaviors where I have to send commands to vehicles multiple times. The commands I'm testing with most are, "arm", changing flight mode, and "takeoff".

I'm running multiple vehicles in SITL. And it appears that the problem gets worse as the number of vehicles and load gets higher. Sometime it can take clicking in the UI 10+ times to send a command before it's received. Clicking and sliding arm, that many times is quite frustrating. 

I'm watching the mavproxy.py console and can see when commands fail, such as arming when it's not ready. But this problem is not that the command is being refused, but that the command is not even getting there to be refused. 

It sometimes seems like qgroundcontrol might be in the middle of doing other operations such as communicating the mission waypoints. 

Here's a video of the behavior with the consoles of all the mavproxy on the left and about 5 vehicles.
https://youtu.be/xP6CEFyIRsU

Once a command is received by mavproxy you can see it print in the console. If the command comes in and is invalid, mavproxy will display the error in the console. Unfortunately I didn't record an example of that. 

I may have some time to look into this but could use some suggestions for debugging. Is there a way to turn on a more verbose output related to communications?

As mentioned this is happening in SITL and on loopback or ethernet devices so I believe that the bandwidth should not be an issue.

**Comment by DonLakeFlyer** [comment]
First thing can you make sure you can reproduce it with a daily build? The Stable build is getting a bit old now. Also does it reproduce without mavproxy connected?

**Comment by DonLakeFlyer** [comment]
Wait is that the mavproxy that comes up with ArduPilot SITL?

**Comment by DonLakeFlyer** [comment]
@tfoote I assume you are using ArduPilot SITL. Can you explain how you are running multilpe ArduPilot SITL instances so I can repro?

**Comment by DonLakeFlyer** [comment]
Also just FYI: The codepath to get these messages out is pretty straightforward. It they really don't make it out the pipe then it is most likely a UDPLink problem. We have had problems with UDP links losing data in the past. Also currently sent commands are not resent if an ack doesn't come back for them. That needs to be changed as well. But that said over a SITL udp link packets should not be lost.

**Comment by tfoote** [comment]
Yes, it's using ardupilot's SITL mavproxy setup. I just got pointed to the possibility of removing mavproxy from the loop: https://github.com/osrf/uctf/issues/50

There's instructions to launch multiple SITL's in this repo: https://github.com/osrf/uctf

I'm actively developing on the payload_v2 branch. 

If you have a xenial machine we have debs that you can dowload and run quickly following these instructions: https://github.com/osrf/uctf/tree/payload_v2/doc/install_binary  


This branch is under active development and is using the ardupilot by default and there's quite a bit of extra stuff being brought up on top that's only partially functional, but shouldn't interfere with the basic ground control interface to the drone. 

After that you can just 
With the environment setup you can
`roslaunch uctf uctf.launch gui:=false` and then in another window with the environment setup
`spawn_blue 1 2 3 4` to get 4 arducopters. Adding '--px4` will switch it to rotors SITL instances instead. I haven't tested the px4 version recently. 

And if you use blue, all the SITL instances can be connected to by the ground control on port 14000

**Comment by DonLakeFlyer** [comment]
Wow, that's a pretty involved toolchain! I'd guess lots of possible failure points unrelated to QGC doing something wrong. I'm going to try to repro this with a direct path to multiple SITL instances.

**Comment by DonLakeFlyer** [comment]
I can repro this just using a direct connection the 5 ArduPilot instances

**Comment by tfoote** [comment]
great, let me know if I can help or test anything

**Comment by DonLakeFlyer** [comment]
Likely same problem as #4235 MAVProxy issue.

**Comment by DonLakeFlyer** [comment]
This is a problem with MAVProxy not sending commands to the right vehicle so they get lost. I've been able to get commands to go through to 3 different vehicels without problem by loading SITL directly without MAVProxy. For some reason anything beyond 3 fails to connect, but that also seems to be some sort of SITL vehicle problem as well. So nothing really going wrong here on the QGC side of things.

I'm going to keep this open since I seem to have lost the Issue about checking for command ack and retry. This will make the above problem a little better. But it will still happen.

**Comment by DonLakeFlyer** [comment]
FYI: The problem with the 4th and 5th vehicle were setup issues. I can run 5 vehicles without problem.

**Comment by DonLakeFlyer** [comment]
Command ack issue is #2975. Will fix in 3.2

**Comment by OskarZyg** [comment]
I'm having a similar issue 10 years later apparently, QGroundControl refuses to send anything over UDP. Very strange behaviour. Only one packet of size 19 every once in a while, presumably a heartbeat.

<details>
<summary>extract from tcpdump of the port</summary>

```
19:55:09.547882 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 36
19:55:09.556032 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 17
19:55:09.558728 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 40
19:55:09.562898 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 45
19:55:09.562906 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 10
19:55:09.567320 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 36
19:55:09.568506 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 22
19:55:09.572708 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 39
19:55:09.574028 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 28
19:55:09.576238 eth0  In  IP 192.168.2.1.62849 > 192.168.2.4.5600: UDP, length 19
19:55:09.577853 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 44
19:55:09.614535 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 36
19:55:09.617462 eth0  In  IP 192.168.2.1.62849 > 192.168.2.4.5600: UDP, length 19
19:55:09.629630 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 40
19:55:09.656785 eth0  In  IP 192.168.2.1.62849 > 192.168.2.4.5600: UDP, length 19
19:55:09.682789 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 36
19:55:09.697604 eth0  In  IP 192.168.2.1.62849 > 192.168.2.4.5600: UDP, length 19
19:55:09.709822 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 38
19:55:09.724903 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 40
19:55:09.736788 eth0  In  IP 192.168.2.1.62849 > 192.168.2.4.5600: UDP, length 19
19:55:09.748465 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 36
19:55:09.760222 eth0  In  IP 192.168.2.1.62849 > 192.168.2.4.5600: UDP, length 17
19:55:09.777952 eth0  In  IP 192.168.2.1.62849 > 192.168.2.4.5600: UDP, length 19
19:55:09.808442 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 28
19:55:09.814553 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 36
19:55:09.827888 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 40
19:55:09.836666 eth0  In  IP 192.168.2.1.62849 > 192.168.2.4.5600: UDP, length 19
19:55:09.859162 eth0  In  IP 192.168.2.1.62849 > 192.168.2.4.5600: UDP, length 19
19:55:09.883608 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 36
19:55:09.897047 eth0  In  IP 192.168.2.1.62849 > 192.168.2.4.5600: UDP, length 19
19:55:09.904683 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 38
19:55:09.925860 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 40
19:55:09.949484 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 36
19:55:09.957282 eth0  In  IP 192.168.2.1.62849 > 192.168.2.4.5600: UDP, length 19
19:55:09.977771 eth0  In  IP 192.168.2.1.62849 > 192.168.2.4.5600: UDP, length 19
19:55:10.015507 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 36
19:55:10.024090 eth0  In  IP 192.168.2.1.62849 > 192.168.2.4.5600: UDP, length 19
19:55:10.024741 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 40
19:55:10.055617 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 45
19:55:10.056666 eth0  In  IP 192.168.2.1.62849 > 192.168.2.4.5600: UDP, length 19
19:55:10.062166 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 28
19:55:10.081442 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 36
19:55:10.097532 eth0  In  IP 192.168.2.1.62849 > 192.168.2.4.5600: UDP, length 19
19:55:10.108909 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 38
19:55:10.126781 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 40
19:55:10.137475 eth0  In  IP 192.168.2.1.62849 > 192.168.2.4.5600: UDP, length 19
19:55:10.150883 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 36
19:55:10.176127 eth0  In  IP 192.168.2.1.62849 > 192.168.2.4.5600: UDP, length 19
19:55:10.217486 eth0  In  IP 192.168.2.1.62849 > 192.168.2.4.5600: UDP, length 19
19:55:10.219384 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 36
19:55:10.226048 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 40
19:55:10.256358 eth0  In  IP 192.168.2.1.62849 > 192.168.2.4.5600: UDP, length 19
19:55:10.259351 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 10
19:55:10.286577 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 36
19:55:10.297837 eth0  In  IP 192.168.2.1.62849 > 192.168.2.4.5600: UDP, length 19
19:55:10.303975 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 38
19:55:10.311939 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 28
19:55:10.325014 eth0  Out IP 192.168.2.4.5600 > 192.168.2.1.62849: UDP, length 40
^C
249 packets captured
253 packets received by filter
0 packets dropped by kernel
```

</details>
